###########################################################
#                                                         #
# Copyright (c)                                           #
#                                                         #
# The Verifiable & Control-Theoretic Robotics (VECTR) Lab #
# University of California, Los Angeles                   #
#                                                         #
# Authors: Kenny J. Chen, Ryan Nemiroff, Brett T. Lopez   #
# Contact: {kennyjchen, ryguyn, btlopez}@ucla.edu         #
#                                                         #
###########################################################

dlio:
  version: piluohong
  loopflag: true
  saveFile: false
  matchMethod: GICP

  adaptive: true

  pointcloud:
    deskew: false
    voxelize: true

  imu:
    calibration: true
    intrinsics:
      accel:
        bias: [ 0.0, 0.0, 0.0 ]
        sm:   [ 1.,  0.,  0.,
                0.,  1.,  0.,
                0.,  0.,  1. ]
      gyro:
        bias: [ 0.0, 0.0, 0.0 ]

  extrinsics:
    baselink2imu:
      t: [-8.086759e-01, 3.195559e-01, -7.997231e-01]
      R: [9.999976e-01, 7.553071e-04, -2.035826e-03, -7.854027e-04, 9.998898e-01, -1.482298e-02, 2.024406e-03, 1.482454e-02, 9.998881e-01]
    baselink2lidar:
      t: [ 0.,  0.,  0. ]
      R: [ 1.,  0.,  0.,
           0.,  1.,  0.,
           0.,  0.,  1. ]

  frames:
    odom: map
    baselink: base_link
    lidar: lidar
    imu: imu

  map:
    waitUntilMove: false
    dense:
      filtered: true
    sparse:
      frequency: 1.0
      leafSize: 0.25

  odom:

    gravity: 9.80665

    imu:
      approximateGravity: true
      calibration:
        gyro: true
        accel: true
        time: 5
      bufferSize: 5000

    preprocessing:
      cropBoxFilter:
        size: 1.0
      voxelFilter:
        res: 0.3

    keyframe:
      threshD: 1.0
      threshR: 30.0
      threshJaccardCorr: 2.0 
      threshJaccardSim: 0.1
      envir_thresh_: 2.0
      loop_score_: 0.3
      loop_search_R: 15
      loop_search_Num: 20
      loop_size: 0.2
      loop_frequeency: 2
      downsamplerate: 1.0

    submap:
      useJaccard: false
      keyframe:
        knn: 20
        kcv: 5
        kcc: 5
    gicp:
      minNumPoints: 2000
      kCorrespondences: 20
      maxCorrespondenceDistance: 1.0
      maxIterations: 100
      transformationEpsilon: 0.005
      rotationEpsilon: 0.005
      initLambdaFactor: 1e-9
      neighborSearch: 1
      dw_dist: 0.00
      dw_azimuth: 0.00
      dw_elevation: 0.00

    geo:
      Kp: 4.5
      Kv: 11.25
      Kq: 4.0
      Kab: 2.25
      Kgb: 1.0
      abias_max: 5.0
      gbias_max: 0.5
